apis/eventstream/index.yaml

```diff
@@ -34,12 +33,30 @@ get:
             type: string
             description: https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events#Event_stream_format
           examples:
-            head:
-              description: The node has finished processing, resulting in a new head
+            state:
+              description: |
+                The node has updated its state.  This event will be sent in three situations:
+
+                - a node has received a block, and updated state accordingly
+                - a node has _not_ received a block four seconds after the start of the slot, and updated state accordingly
```

Comment:
Is this per spec? I don't think we should include this, as it depends on how each client handles this. 

It makes a lot more sense to introduce separate clock:slot/slot event so vc can listen on start of beacon chain slot.
We had some issues where VC clock is few microseconds ahead of Bn clock so it asks for duties and other actions a bit prematurely.

Reply:
There was quite a lengthy discussion on this in the Discord starting at https://discordapp.com/channels/595666850260713488/710831784991916072/760955856165797890 but to focus on the first point of `chain_head` Vs. `state`: the discussion was around the fact that state can change even though the chan head has not (_e.g._ in the case of skipped slots).  In this situation, clients need to be made aware that state has updated hence the move from head to state.

The other point of the `state` event is to subsume the requirement for clock events.  The clock event basically says "I'm ready to start serving for state _x_", so we're providing much the same information as we already do in the state event.  However, if an implementation is not ready to serve for state _x_ as soon as that slot starts (for example, if it is carrying out internal accounting for the epoch transition) then sending a clock event has no value.  I think we're talking about much the same thing with clock and state events, it's just that the state event covers instances where beacon nodes are not immediately ready at a given clock tick.

One thing:

> ...vc can listen on start of beacon chain slot and decide on it's own how long it wants to wait.

A major advantage of the state event is that the VC doesn't have to do this guess; as soon as it receives the state event it knows it is good to query the BN for information related to that state.

Reply:
> the discussion was around the fact that state can change even though the chan head has not (e.g. in the case of skipped slots)

But it cannot really. that's just client optimization that may or may or may not payoff? 
There could be block even though client didn't receive it so client might do useless epoch transition which not every client can afford :smile: 

> The clock event basically says "I'm ready to start serving for state x"

First chain_head is marking that not slot event, I don't think client should calculate states ahead of chain_head and returning it, that should be error 404.
Second, we can add doc saying that this just marks start of slot in BN clock.

Reply:
Perhaps worth going through a specific situation in more detail.

The first functional sentence of the validator flow doc is "On start of every epoch, validator should fetch proposer duties.".  Given we cannot assume that the validator client and beacon node clocks are in sync, the beacon node should indicate to the validator client when it thinks the epoch has started so that the validator client can fetch the proposer duties.  This is one of the clock events that have been discussed.

However, a simple clock event is not enough.  This is because the some beacon nodes do not update their state until the epoch starts.  This means that there is a time after the start of the epoch when the beacon node will return the previous epoch's information.  So we need some sort of event to say not just "the epoch has started" but "I'm ready to serve data for the epoch".  For some beacon node implementations this may be the same thing, but for others it may not.  Either way, though, we can use a single `state` event to notify clients that the beacon node is ready to serve.

Does this help to clarify where the `state` event would be used, and how it combines features of the current `head` and discussed `clock` events?

(We can discuss the other features of the `state` event separately, but I wanted to try to lay some common ground rather than attempt to tackle all of the different points at once).

Reply:
Using 4 seconds here makes two assumptions:

- That 4 seconds is relevant to the user (it's the case if they're producing attestations, but might not always be the case)
- The client is able to perform the required skip-slots within 4s.